### ISSUE TITLE
Prevent `redocly bundle` from creating new schema objects in `#/components/schemas` when using `--dereferenced` option

### ISSUE THREAD

**Issue by outdooracorn** [Redocly/redocly-cli#1798]
**Describe the bug**

The `redocly bundle` command is adding schema objects to the `#/components/schemas` section of the bundled schema, even when using the `--dereferenced` option is used.

Closed issue #239 seems to be similar and points to using the `--dereferenced` option to solve this issue.

**To Reproduce**

1. [Given this `redocly.yaml` file](https://github.com/outdooracorn/redocly-cli-dereference-example/blob/577cb72475548464182491423993b8d422b22c7f/redocly.yaml) 
2. [And these OpenAPI files](https://github.com/outdooracorn/redocly-cli-dereference-example/tree/577cb72475548464182491423993b8d422b22c7f/schema)
3. Run `npx redocly bundle schema/openapi.json --dereferenced -o out/openapi.json`
4. See that `PropertyValuePair` and `Error` schema objects are added to `#/components/schemas` in the [bundled OpenAPI document](https://github.com/outdooracorn/redocly-cli-dereference-example/blob/577cb72475548464182491423993b8d422b22c7f/out/openapi.json)

**Expected behavior**

I would expect only (dereferenced versions of) the schema objects present in [`schemas.json`](https://github.com/outdooracorn/redocly-cli-dereference-example/blob/577cb72475548464182491423993b8d422b22c7f/schema/schemas.json) to be in the `#/components/schemas` section of the bundled and dereference OpenAPI document.

**Logs**

**OpenAPI description**

OpenAPI v3.1 - see [example repo](https://github.com/outdooracorn/redocly-cli-dereference-example/tree/577cb72475548464182491423993b8d422b22c7f)

**Redocly Version(s)**

`"@redocly/cli": "^1.25.11",`

**Node.js Version(s)**

`v18.20.4` and `v20.12.2`

**OS, environment**

Kubuntu 22.04

**Additional context**

**Comment by tatomyr** [comment]
Have you tried using the `--remove-unused-components` [option](https://redocly.com/docs/cli/decorators/remove-unused-components#remove-unused-components)?

**Comment by outdooracorn** [comment]
> Have you tried using the `--remove-unused-components` [option](https://redocly.com/docs/cli/decorators/remove-unused-components#remove-unused-components)?

I have. It removes some, but not all, of the components. In this example, `#/components/schemas/Statement` is left.

I also don't want to remove all the components. I would like to keep `'#/components/schema` so that it displays the schemas section in Swagger UI.

**Comment by tatomyr** [comment]
Ah, so you want to bundle your files, preserve the existing components, and not add the new ones? Then it's more of an enhancement rather than a bug. I'm not sure it would be an easy change. It will require modifying the internals, as the bundler currently resolves external references into the `components` section regardless of the `--dereferenced` option because the component might be used in an indirect way (e.g. in a discriminator). 
However, you can try using a [custom plugin](https://redocly.com/docs/cli/custom-plugins/custom-decorators) with a decorator to filter schemas that were only present in the original components. It should probably look like this: 

```js

const filterSchemas = () => {
  let originalSchemas;
  return {
    Components: {
      leave(components) {
        for (const key in components.schemas || {}) {
          if (originalSchemas?.[key] === undefined) {
            delete components.schemas[key];
          }
        }
      },
      NamedSchemas: {
        enter(schemas) {
          originalSchemas = schemas;
        },
      },
    },
  };
};
```

Please let me know if that works!

**Comment by outdooracorn** [comment]
Ideally, we would like to bundle the files and only dereference/inline schemas that aren't already in `#/components`. Said differently, keep references to components already in `#/components`, and dereference/inline everything else.

Thanks for the custom decorator plugin code, we will try it out.

**Comment by tatomyr** [comment]
Hi @gnvk. I don't see an easy workaround for your problem. To achieve that, we'd need to change the current behaviour of Redocly CLI which might impact existing users. Have you considered renaming the duplicated files into something like `StockTrade.yaml` or `crypto-trade.yaml`, so you'll get the corresponding names in the `components`?

**Comment by gnvk** [comment]
> Have you considered renaming the duplicated files into something like StockTrade.yaml or crypto-trade.yaml, so you'll get the corresponding names in the components?

Yes, in fact I even did all the renames, just to realise that that messes up the generated files for the server components. For example the trade in `stock-trade.yaml` will become `stock.StockTrade` in Go, which is something we don't want. 

Basically, I can either generate the correct `openapi.yaml` by renaming the component files, as you suggested, or I can generate the correct "small" YAMLs (e.g. `stock.yaml`) by leaving them as they're now. With swagger-cli I can achieve both using the named schemas, which are currently ignored by redocly. This is why I feel this issue is a blocker for us.

**Comment by tatomyr** [comment]
@gnvk  I see. Then the only workaround I can suggest is running a decorator on the stock.yaml to remove the needless `stock-` prefixes in components and refs. It seems to be quite fragile though.

Just in case you want to explore that path, here's the idea. First, you have to create a plugin like the following:

```js
// my-plugin.mjs

export default () => {
  return {
    id: "my-plugin",
    decorators: {
      oas3: {
        "remove-prefix": ({ prefix }) => ({
          ref: {
            leave(refObject) {
              // after partially resolving refs, we check if it still has a $ref field
              if (refObject?.$ref) {
                // ...and if it does, we remove the prefix
                refObject.$ref = refObject.$ref.replace(prefix, "");
              }
            },
          },
          NamedSchemas: {
            leave(schemas) {
              // replace all keys that start with the prefix with the new key in components.schemas
              for (const key in schemas) {
                if (key.startsWith(prefix)) {
                  const newKey = key.replace(prefix, "");
                  schemas[newKey] = schemas[key];
                  delete schemas[key];
                }
              }
            },
          },
        }),
      },
    },
  };
};
``` 

Then, use it use it in you configuration file:

```yaml
# redocly.yaml

plugins:
  - plugin.mjs

apis:
  stock:
    root: api/stock.yaml
    output: generated/redocly/stock.yaml
    # set up the custom decorator specific to the 'stock' api
    decorators: 
      "my-plugin/remove-prefix": 
        prefix: "stock-"
  openapi:
    root: api/openapi.yaml
    output: generated/redocly/openapi.yaml
```

You don't need to declare the `components` section in your API files anymore but you have to rename the files to contain the prefixes (like `stock-trade.yaml` &c.) which will be removed by the `remove-prefix` decorator.

This configuration will produce the same generated openapi.yaml (`redocly bundle openapi`) but it will update the component schema names that start with the `stock-` prefix when generating the stock.yaml (`redocly bundle stock`).

Please let me know it that helps.

**Comment by gnvk** [comment]
Thank you @tatomyr! With a tiny modification in the plugin script (we also need to replace the `NamedParameters`) your workaround seems to be working on my [example](https://github.com/gnvk/redocly_example/pull/1). On the other hand, the same change on our actual codebase is a massive task. I started it, so far it looks good...
